opm/simulators/linalg/cuistl/CuBuffer.cpp

```diff
@@ -0,0 +1,227 @@
+/*
+  Copyright 2022-2023 SINTEF AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <algorithm>
+#include <fmt/core.h>
+#include <opm/simulators/linalg/cuistl/CuBuffer.hpp>
+#include <opm/simulators/linalg/cuistl/CuView.hpp>
+#include <opm/simulators/linalg/cuistl/detail/cuda_safe_call.hpp>
+#include <opm/simulators/linalg/cuistl/detail/vector_operations.hpp>
+
+namespace Opm::cuistl
+{
+
+template <class T>
+CuBuffer<T>::CuBuffer(const std::vector<T>& data)
+    : CuBuffer(data.data(), detail::to_int(data.size()))
+{
+}
+
+template <class T>
+CuBuffer<T>::CuBuffer(const size_t numberOfElements)
+    : m_numberOfElements(detail::to_int(numberOfElements))
+{
+    OPM_CUDA_SAFE_CALL(cudaMalloc(&m_dataOnDevice, sizeof(T) * detail::to_size_t(m_numberOfElements)));
+}
+
+template <class T>
+CuBuffer<T>::CuBuffer()
+    : m_numberOfElements(0)
+{
+    OPM_CUDA_SAFE_CALL(cudaMalloc(&m_dataOnDevice, 0));
```

Comment:
This is weird... What happens when you allocate 0 bytes? Is it not better to just keep m_dataOnDevice a null pointer?

Reply:
Just scratch this base case constructor? Local tests indicate that this works, but I agree it looks dubious.